umh-core/pkg/service/redpanda/redpanda.go

```diff
@@ -956,3 +962,181 @@ func formatMemory(memory int) string {
+	if err != nil {
+		return fmt.Errorf("failed to create readback request: %w", err)
+	}
+
+	readbackResp, err := s.httpClient.Do(readbackReq)
+	if err != nil {
+		// If we get a connection refused error, it means that the Redpanda service is not yet ready (or is in the progress of restarting), so we just ignore it
+		if strings.Contains(err.Error(), "connection refused") {
+			return nil
+		}
+		return fmt.Errorf("failed to send readback request: %w", err)
+	}
+
+	readbackBody, err := io.ReadAll(readbackResp.Body)
+	if err != nil {
+		return fmt.Errorf("failed to read readback response body: %w", err)
+	}
+
+	err = readbackResp.Body.Close()
+	if err != nil {
+		return fmt.Errorf("failed to close readback response body: %w", err)
+	}
+
+	// Parse as JSON
+	var readbackConfig map[string]interface{}
+	if err := json.Unmarshal(readbackBody, &readbackConfig); err != nil {
+		return fmt.Errorf("failed to unmarshal readback response body: %w", err)
+	}
+
+	// Validate all keys against the readback config
+	// Since the types of our request and the response might differ we need to handle the different types
+	for key, value := range configUpdates {
+		readbackValue, ok := readbackConfig[key]
+		if !ok {
+			return fmt.Errorf("key %s not found in Redpanda cluster config for %s", key, redpandaName)
+		}
+
+		// If readback value is a string, we can just compare the strings
+		if readbackValueString, ok := readbackValue.(string); ok {
+			if readbackValueString != value {
+				return fmt.Errorf("value of %s in Redpanda cluster config for %s is not the same as the value in the request [string comparison]", key, redpandaName)
+			}
+		}
+		// Otherwise we can assume that the value is a float64, and we can therefore compare the float64s
+		readbackValueFloat, ok := readbackValue.(float64)
+		if !ok {
+			return fmt.Errorf("value of %s in Redpanda cluster config for %s is not a float64 [readback]", key, redpandaName)
+		}
+		valueFloat, ok := value.(float64)
+		if !ok {
+			// Try to cast an int and then to a float64
+			valueInt, ok := value.(int64)
+			if !ok {
+				return fmt.Errorf("value of %s in Redpanda cluster config for %s is not a float64 [request]", key, redpandaName)
+			}
+			valueFloat = float64(valueInt)
+		}
+
+		if readbackValueFloat != valueFloat {
+			return fmt.Errorf("value of %s in Redpanda cluster config for %s is not the same as the value in the request [float64 comparison]", key, redpandaName)
+		}
+	}
+
+	return nil
+}
+
+// UpdateRedpandaClusterConfig updates Redpanda's cluster configuration through its admin API.
+// The function performs two key operations:
+//  1. Sends a PUT request to update the cluster configuration
+//  2. Verifies the changes by reading back and comparing the new values
+//
+// Note on restarts:
+//   - This function does not directly trigger a Redpanda restart
+//   - Restarts are handled by the reconcile loop when S6 service config changes
```

Comment:
what does happen first? the s6 service change, or the API call? ideally it should always do the API call first, but could there also be cases where redpanda is restarted and then the API call is attempted and fails?

Reply:
good catch